# #1 - Machine Learning Project :: Wine Style Prediction
---
• In this project I will create **a model to predict the type of wine (red or white)**, using a database of almost **6,500 rows** and **13 columns** with varied information about wine characteristics such as : density, ph, alcohol, quality, acidity etc.

• The database used was from Kaggle: https://www.kaggle.com/datasets/dell4010/wine-dataset?resource=download

• The video used to reference in the construction of the model: https://www.youtube.com/watch?v=DNegkxwQJuY

## 1. Library Imports and Database Tuning

In [1]:
import pandas as pd

In [2]:
dataset = pd.read_csv('dataset/wine_dataset.csv')
rows, columns = dataset.shape
print(f'The database have: {rows} rows and {columns} columns')
dataset.head() # sample of dataset

The database have: 6497 rows and 13 columns


fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality style  
0      9.4        5   red  
1      9.8        5   red  
2      9.8        5   red  
3      9.8        6   red  
4      9.4        5   red

Checking for missing values

In [3]:
dataset.isnull().sum().sort_values(ascending=False) # there's no one 'NaN' value

fixed_acidity           0
volatile_acidity        0
citric_acid             0
residual_sugar          0
chlorides               0
free_sulfur_dioxide     0
total_sulfur_dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
style                   0
dtype: int64

In [4]:
# Styles distribution in dataset
dataset['style'].value_counts()

white    4898
red      1599
Name: style, dtype: int64

In [5]:
# Turning styles strings into binary numbers to modeling
dataset['style'] = dataset['style'].replace('red', 0)
dataset['style'] = dataset['style'].replace('white', 1)

In [6]:
# Checking the styles distribution
dataset['style'].value_counts()

1    4898
0    1599
Name: style, dtype: int64

## 2. Preparing the Prediction Model
---
The decision tree model was selected to make the prediction because is an algorithm used for classifications

Separating predictor variables and target variables:

In [7]:
x = dataset.drop(columns=['style']) # Predictor vars
y = dataset['style'] # Target vars

The test sample size was defined as **30%** ("test_size=0.3").

In [8]:
from sklearn.model_selection import train_test_split

# Separating data to train and test
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3)

In [9]:
from sklearn.ensemble import ExtraTreesClassifier # Model used to prediction

model = ExtraTreesClassifier()
model.fit(x_train, y_train)

result = model.score(x_test, y_test)
print(f"Accuracy: {result:.2%}")

Accuracy: 99.79%


In [10]:
prediction = model.predict(x_test)

In [11]:
# Validation using Confusion Matrix
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, prediction)

array([[ 479,    4],
       [   0, 1467]], dtype=int64)

## 3. Conclusion
---
**Accuracy**: Despite the high accuracy of 99.79%, it is necessary to be aware of data bias, which may indicate false accuracy and, therefore, an inefficient model. However, as this dataset is a simple and validated structure, it can be understood as a high performance model, since it deals with more homogeneous and compact data, with a low number of predictive variables.